Add a chapter on Docker

rrtheonlyone opened the pull request:
Fixes #86

This is still a work-in-progress! (Although ready to be peer-reviewed).

Still have to look into:

- [x] Improving Flow
- [x] More diagrams or explanations
- [x] Clear Chapters (rather than information overload)

rrtheonlyone committed:
Add draft book chapter

rrtheonlyone commented:
@monmanuela @junming403 Do have a look when you guys are free :D

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+
+#### 1.1 Virtual Machine (VM)
+
+Before the rise of Docker, people used VMs. A VM is an emulation of a real computer that executes program like a real computer. A VM is used to isolate an application and its dependancies into a self-contained unit that can run anywhere.
```
"... that executes _programs_ ..."

monmanuela replied:
People used VMs to do what?

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+
+Before the rise of Docker, people used VMs. A VM is an emulation of a real computer that executes program like a real computer. A VM is used to isolate an application and its dependancies into a self-contained unit that can run anywhere.
+
+![alt text](./Virtual_Machine.png "VM Diagram")
```
Include the source of the diagram since I assume you didn't draw this yourself. Same goes to the other diagram below.

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+
+#### 1.2 Container
+
+As mentioned above, Docker uses Containers, not VMs. The one big difference between containers and VMs is that containers *share* the host system’s kernel with other containers 
```
Missed a full stop at the end

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+
+![alt text](./Container.png "Container Diagram")
+
+This diagram shows you that containers package up just the user space, and not
```
What is "the user space"? Clarify that this refers to the Bins/Libs and App (if I understood this correctly)

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+
+With reference to the above diagram, VMs run on top of a physical machine using a “hypervisor”. A VM is able to provide virtual hardware, a kernel (i.e. OS) and user space. A Windows user can use a VM to run an application that is only supported on OSX for example. 
+
+#### 1.2 Container
```
Just an idea, maybe at the end of the `1.2 Container` section, you can include a table of side by side comparison between VMs and Container as a short summary to show the differences.

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+### 2. Why learn Docker?
+
+Docker is a powerful tool that is rapidly gaining popularity. Many leading companies (e.g. Spotify, Nginx, ElasticSearch etc.) are using Docker for their deployment.
+Here are some of the many advantages that docker provides:
```
Capitalize "docker"

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+- **Rapid application deployment** – containers include the minimal runtime requirements of the application, reducing their size and allowing them to be deployed quickly.
+- **Portability across machines** – an application and all its dependencies can be bundled into a single container that is independent from the host version of Linux kernel, platform distribution, or deployment model. This container can be transfered to another machine that runs Docker, and executed there without compatibility issues.
+- **Version control and component reuse** – you can track successive versions of a container, inspect differences, or roll-back to previous versions. Containers reuse components from the preceding layers, which makes them noticeably lightweight.
+- **Sharing** – you can use a remote repository to share your container with others.Red Hat provides a registry for this purpose, and it is also possible to configure your own private repository.
```
Insert space before the 2nd sentence

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+Docker is a powerful tool that is rapidly gaining popularity. Many leading companies (e.g. Spotify, Nginx, ElasticSearch etc.) are using Docker for their deployment.
+Here are some of the many advantages that docker provides:
+
+- **Rapid application deployment** – containers include the minimal runtime requirements of the application, reducing their size and allowing them to be deployed quickly.
```
Just a minor nit, you should probably capitalize the first words here like "Containers include ..." because it's the start of the sentence. But up to you.

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+
+Let us take a detailed look at some of the above advantages.
+
+#### 2.1 Docker is lightweight 
```
It's quite weird to see that the advantages of Docker (like Docker is lightweight) is put as a subsection of "Why learn Docker". I think it's better if it's a new section by itself: "3. Advantages of Docker"

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+container. 
+
+When you update configurations or install applications in your container, the
+corresponding docker image gets updated as well. This image can then be shared
```
Capitalize Docker here and in other places where you use the word

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+
+When you update configurations or install applications in your container, the
+corresponding docker image gets updated as well. This image can then be shared
+across multiple machines by different people! This way you dont have to
```
The use of the exclamation mark here is quite weird.

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+difficult to run an Docker file that someone else has built. Many of the docker
+hub images also provide sample docker file builds!
+
+On top of all this, Docker has excellent documentation on their website and
```
You can link the documentation here

monmanuela commented on contents/docker/Docker.md:
Just "3" not "3.0"

monmanuela reviewed (commented):
Overall, I find this quite a well-written article on Docker! As someone who isn't very familiar with the topic, I could follow and understand what you're writing. Just some nits, but it looks good to go.

junming403 commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+This diagram shows you that containers package up just the user space, and not
+the kernel or virtual hardware like a VM does. Each container gets its own
+isolated user space to allow multiple containers to run on a single host
+machine. The operating system level architecture is being shared across containers. 
```
After introducing container and VM, probably have a small section telling user explicitly what are the benefits of container over VMs(More efficient space usage, less performance overhead?)

junming403 commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+![alt text](./Virtual_Machine.png "VM Diagram")
+
+With reference to the above diagram, VMs run on top of a physical machine using a “hypervisor”. A VM is able to provide virtual hardware, a kernel (i.e. OS) and user space. A Windows user can use a VM to run an application that is only supported on OSX for example. 
+
```
This is a very general introduction to container, people not familiar with OS might find it hard to understand(For example, How does it packages up user space?) Probably provides some links for further study of how container works(Liked what I have presented). Probably mention that docker utilizes OS-level support(namespaces, cgroups) and provides link to it

junming403 commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+There are plenty of guides out there that show you the extended functionality a
+docker file can provide. Even if you are a complete beginner, it is not
+difficult to run an Docker file that someone else has built. Many of the docker
+hub images also provide sample docker file builds!
```
Actually I think docker-hub community is the key reason why docker is so easy to get started with, because in most of the cases, you do not need to build your own image! And even if you do, you can just extend existing images. I feel this point should be emphasised.

junming403 commented on contents/docker/Docker.md:
Can add a resource on how docker works, as I think it is important to know how docker is implemented so as to know when it appropriate to use docker etc. 
For example http://docker-saigon.github.io/post/Docker-Internals/

junming403 reviewed (commented):
Overall good works! I have added some comments for discussion. 
Another worth mention points of docker: It really makes local development environment much cleaner since all the dependencies are inside the container. As such, it also makes set up local development much easier, that is the reason I feel docker is really suitable for open source projects because it makes it much easier for contributors to get started without worrying about it will mess up the local environment.

damithc commented:
Some quick comments:
* Also add a link to this article from the main landing page
* Most other articles in the book don't use heading numbers. Remove heading numbers to be consistent?
* Images are too big. Resize until their text size is consistent with the text size of the main text.
* Current: what is VM? how containers is different from VM?
  An alternative: what is a container? how is it different from VM? 
  Not sure which one is better.

ongspxm commented on contents/docker/Docker.md:
mention something about `compose`? like how docker allows you to setup multiple things at the same time?

ongspxm commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,161 @@
+As we have discussed, Docker uses makes use of Containers that share the host
+system's kernel with other containers. It runs a discrete process, taking no more memory than any other executable, making it lightweight.  
+
+Unlike virtual machines, containers do not have the high overhead and hence enable more efficient usage of the underlying system and resources.
```
a little repetitive, like having 2 conclusions, combine?

rrtheonlyone committed:
Improve book chapter

rrtheonlyone committed:
Add Docker to nav bar (under tools

rrtheonlyone commented:
Thanks for the review everyone.

I have done an overhaul of my current article and rewritten it based off the feedback. I have also added Docker under `Tools Useful for Software Engineers`

The preview is available [here](https://deploy-preview-96--learningresources.netlify.com/contents/docker/docker).

Please give me another round of review when you have the time.

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,155 @@
+
+With the increasing complexity of these applications, deployment starts to become a concern. Instead of running one service, some applications will require you to run a few services at the same time. There also needs to be configurations setup for each of these services. Applications like Docker help to make deployment a seamless process. This article provides an introduction to Docker along with ...
+
+### 1. What is Docker?
```
Remove the number? (Since the other sub-headings don't follow the numbering)

rrtheonlyone replied:
Good catch. Yup will remove

monmanuela commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,155 @@
+
+The set-up to use `docker-compose` involves the creation of a YAML file. Detailed information on this is available [here](https://docs.docker.com/compose/overview/).
+
+### When to use docker?
```
Capitalize 'Docker'

rrtheonlyone replied:
Done

monmanuela reviewed (changes requested):
Just a nit besides the comment: perhaps at the end put a short 'Conclusion' and 'Further Reading'. At the moment the article looks good, just that the ending feels a bit abrupt.

rrtheonlyone committed:
Make changes according to PR comments

rrtheonlyone commented:
Updated according to review comments

junming403 commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+<box type="danger">
+As you can read from above, Docker is undeniably a powerful tool that can alleviate your deployment troubles. However, be aware that Docker is not a one-size-fits-all solution to your problems. Docker also has its limitations e.g. refer to this <a href="https://www.channelfutures.com/open-source/when-not-to-use-docker-understanding-the-limitations-of-containers/">article</a>. Carefully consider y ...
+</box>
```
I am not sure if we should warn users about its limitations here, probably put this as a separate section called "limitations" at the end of this article is better?

junming403 commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+The table below lists some of the differences between a VM and a container:
+
+<img src="VMvsContainer.png" alt="VM vs Container" width="100%">
```
I feel a bit duplication here as the difference has already been mentioned above.

junming403 commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+Now that we have a good idea of how Docker works and a summary of the notable advantages it is able to provide, let us take a closer look at some of the unique features that Docker is able to provide.
+
+#### Benefit: Docker is lightweight 
```
It seems that **Benefits** and **why using docker** is a bit overlapping, consider merging these two sections into one? 
I use to use this organization and prof suggested me to reorganize into something like:

1. Wha is docker
2. Why use docker 
3. How to get stared

junming403 reviewed (commented):
A well-written article that I really enjoyed reading,  A small suggestion would be reorganize the article structure as the comment i left, but it's up to you to decide.

rrtheonlyone commented:
@junming403 thanks for the review! 

I am also not very clear on the structure myself. I feel the current one should allow a new reader to get a good understanding of docker. 

@damithc possible to give a preliminary review on this?

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+The idea of containers is not so different from Virtual Machines (VM). Before the rise of Docker, people used VMs. A VM is an emulation of a real computer that is used to isolate an application and its dependencies into a self-contained unit that can run anywhere.  For example, you could use a VM to run an application on a Linux system even though you are running Windows.
+
+However, a container posses some advantages over a VM. With reference to the [Docker Documentation](https://docs.docker.com/get-started/), a container runs natively on Linux and shares the kernel of the host machine with other containers. It runs a discrete process, taking no more memory than any other executable, making it lightweight.
```
`With reference to` -> `As per`?

rrtheonlyone replied:
Done

damithc reviewed (commented):
Structure is generally OK. If anything, you can get rid of the summary of advantages and absorb that into the list of benefits. If you want to give both a summary and a detailed version, it's best to make the two match exactly. i.e., list the advantages with 1-2 sentence description (like a table of contents) and then expand on each benefit as a separate section. Otherwise it may be somewhat confusing to the reader to see benefits described in two different ways that don't match up exactly.

junming403 commented:
LGTM!

junming403 reviewed (approved):
LGTM!

rrtheonlyone committed:
Update phrasing

rrtheonlyone commented:
Ready for final review @damithc 

I have looked through the chapter once more. The way I structured it is as follows: Give a concise, list of all advantages Docker is able to provide followed by a detailed explanation of some of these benefits.

The list of advantages at the top is not a summary of the benefits I have listed below. They actually have more points (compiled from a source) and is meant to give readers a quick overview on docker's benefits without having to read the whole chapter. Interested readers can then read further for more information on some of these advantages.

I can also change it to summarise the points I have below! However, I am not sure if that would be better.

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+A Microservices architecture is when you split your application into a set of smaller, interconnected services that each have a specific role (as opposed to one big application). Read up more about it <a href="https://articles.microservices.com/monolithic-vs-microservices-architecture-5c4848858f59">here.</a>
+</box>
+
+With the increasing complexity of these applications, deployment starts to become a concern. Instead of running one service, some applications will require you to run a few services at the same time. There also needs to be configurations setup for each of these services. Applications like Docker help to make deployment a seamless process. This article provides an introduction to Docker along with ...
```
This could be a nice way to start a presentation about docker (i.e., PUNCH), but I'm not sure if that works for an article. Why bring in an entirely new (and not strictly necessary) topic at the very start? If you remove it, the reader wont feel like something is missing.

rrtheonlyone replied:
Yes you are right! I have removed this.

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+> A container is a standard unit of software that packages up code and all its dependencies so the application runs quickly and reliably from one computing environment to another. <sub>--[Docker Website](https://www.docker.com/resources/what-container)</sub>
+
+The idea of containers is not so different from Virtual Machines (VM). Before the rise of Docker, people used VMs. A VM is an emulation of a real computer that is used to isolate an application and its dependencies into a self-contained unit that can run anywhere.  For example, you could use a VM to run an application on a Linux system even though you are running Windows.
```
Can use italics when a new term is introduced e.g., _Virtual Machine_

rrtheonlyone replied:
Done

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+By contrast, a virtual machine (VM) runs a full-blown “guest” operating system with virtual access to host resources through a hypervisor. In general, VMs provide an environment with more resources than most applications need. The diagram below gives a depiction of this:
+
+<img src="Docker_Architecture.png" alt="Docker Architecture" width="100%">
```
It's best to put the diagram just _above_ the text that refers to it.

rrtheonlyone replied:
Done

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+The idea behind Docker is to spin up a lightweight container that is able to execute services quickly without the overhead of a full-blown VM. With this in mind, let us move onto the the reasons to switch over to Docker. The subtleties between a container and VM will become more clear as you read the next few sections.
+
+### Why learn Docker?
```
Do we need the word `learn` here? It's specific to learning, right?

rrtheonlyone replied:
Yup correct. Updated

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+Now that we have a good idea of how Docker works and a summary of the notable advantages it is able to provide, let us take a closer look at some of the unique features that Docker is able to provide.
+
+#### Benefit: Docker is lightweight 
```
Shouldn't this be `Feature: ...`? If not, the natural question is "what's the difference between advantage and benefit?"

rrtheonlyone replied:
Yup thats true

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+
+<img src="VMvsContainer.png" alt="VM vs Container" width="100%">
+
+<sub>_Figure 4. VM vs Container_ [(source)](https://www.backblaze.com/blog/vm-vs-containers/)</sub>
```
Can consider centering images/captions. Also consider using `<pic>` https://markbind.org/userGuide/usingComponents.html#pictures

rrtheonlyone replied:
Yup have updated all my images to use `<pic>`

damithc commented on contents/docker/Docker.md:
```diff
@@ -0,0 +1,165 @@
+Here are some resources that would aid you in this process:
+- https://docs.docker.com/get-started/
+- https://scotch.io/tutorials/getting-started-with-docker
+- https://medium.freecodecamp.org/a-beginner-friendly-introduction-to-containers-vms-and-docker-79a9e3e119b 
```
Can be a bit more detailed. e.g.,
> * [Docker's own documentation on how to get started](https://docs.docker.com/get-started/) - {Give your own assessment of this resource}

Also, try to make this a _learning path_ (i.e., in your opinion, this is the order in which the reader should go) rather than just a bunch of links.

rrtheonlyone replied:
Yes! Made it into a learning path as suggested

damithc reviewed (changes requested):
No major issues 👍 
Some suggestions added

rrtheonlyone committed:
Improve book chapter

rrtheonlyone commented:
Ready for another review.

I have centralised the images and improved the `How to use Docker?` by providing a learning path instead of a bunch of links.

rrtheonlyone committed:
Improve book chapter

rrtheonlyone committed:
Improve book chapter

damithc committed:
Tweak page top

damithc commented:
Merged and deployed. 
Good work @rrtheonlyone 
Thanks to reviewers too.
